ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -50,11 +50,22 @@ const BIT_ARRAY_SHIFT: usize = BIT_ARRAY_LEN.ilog2() as usize;
 // A smaller value is used for tests, to enable covering some corner cases with a
 // reasonable runtime. Some of these tests use TARGET_PROOF_SIZE directly, so for tests
 // it does need to be a power of two.
+//
+// TARGET_PROOF_SIZE is closely related to MAX_PROOF_RECURSION; see the assertion that
+// `uv_values.len() <= max_uv_values` in `ProofBatch` for more detail.
 #[cfg(test)]
 pub const TARGET_PROOF_SIZE: usize = 8192;
 #[cfg(not(test))]
 pub const TARGET_PROOF_SIZE: usize = 50_000_000;
 
+/// Returns the largest power of two less than or equal to `self`.
+///
+/// Returns 0 if the input is outside the range `1..1 << (usize::BITS - 1)`.
+#[must_use]
+pub const fn prev_power_of_two(value: usize) -> usize {
```

Comment:
there is another version of it [here](https://github.com/private-attribution/ipa/blob/8e61beebad2ed498c3b3b108c581903534a16aaf/ipa-core/src/helpers/gateway/send.rs#L261). 
Maybe we can follow up to merge them